File: Blind75/Arrays&Hashing/src/productExceptSelf.java


public class productExceptSelf {
    public int[] productExceptSelf(int[] nums) {
        
        int[] ans = new int[nums.length];
        int prod=1;
        ans[0] = 1;
        for(int i=1 ; i<nums.length ; i++)
        {
            ans[i] = prod*nums[i-1];
            prod*=nums[i-1];
        }
        int prod2=1;
        int post =1;
        for (int j=nums.length-1; j>0; j--)
        {
            post = nums[j]*prod2;
            ans[j-1] *= post;
            prod2*=nums[j];
        }

        return ans;
    }

}

File: Blind75/Python/Arrays&Hashing/ProductExceptSelf.py
from typing import List


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        res=[1]*len(nums)
        prefix =1
        for i in range(len(nums)-1):
            res[i] = prefix
            prefix*=nums[i]
            
        suffix = 1
        for i in range(len(nums)-1,-1,-1):
            res[i] *= suffix
            suffix *= nums[i]

        return res



if __name__ == "__main__":
    nums = [1, 2, 3, 4]
    sol = Solution()
    result = sol.productExceptSelf(nums)
    print("Input:", nums)
    print("Output:", result)